### PR TITLE
Add CODEOWNERS feature

### LINKED ISSUE #10161: support Github-like CODEOWNERS?
We are a team of ~15 engineers using gitea for daily development.

What is the attitude of this project towards supporting the github-like CODEOWNERS system or something like it?

To be honest, I'd find this functionality very useful.

### PR DESCRIPTION
**Pull request by bcallaghan-et**
Adds the ability to define code ownership in a CODEOWNERS file, similar to the [GitHub feature](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners). See the new docs page for a full feature description.

Closes #10161

### Implementation
The primary back-end change is adding a call to `AddCodeownerReviewers()` in `services/issue/issue.go`. This searches the base repo for a valid CODEOWNERS file, parses it, and requests reviews from any users/teams that own any changed files in the PR. The parsing/validation is contained in functions within `services/issue/codeowners.go`.

The applicable CODEOWNERS file is also parsed to:
* Display validation errors when viewing/editing a CODEOWNERS file
* Display the hoverable shield icon inticating ownership as described below 

### UI Changes

Viewing a CODEOWNERS file with validation errors:
Validation information on view

Editing a CODEOWNERS file with validation errors:
Validation information on edit

The remaining examples assume the following valid CODEOWNERS file is in the `main` branch:

```
asdf/ @user3 @Org1/teamA
*.txt @Org1/teamA
```

Viewing a file that has an owner displays a shield icon that can be hovered over to show the ownership tooltip. If no owner applies to the file, the shield is not displayed:
Shield when viewing a file

When opening a PR, the same hoverable codeowner shield icon is visible for any changed files with owners:
Shields when creating pull request

Same thing when viewing a PR's changed files:
Shields when viewing PR changed files

Same thing when viewing a commit:
Shields when a commit

### Known issues and features we are still working on
* Testing. We recently ran into some issues integrating our unit tests properly.
* Currently anyone can create/edit a CODEOWNERS file. This needs to be restricted to users with admin/owner permissions to match the GitHub feature.
* Codeowners should not be added when a PR is created as WIP, but should then be added when the WIP prefix is removed.
* Some code can be cleaned up and refactored a bit in `codeowners.go`

### CODE REVIEW

**Comment by yardenshoham** [comment]
Duplicate of https://github.com/go-gitea/gitea/pull/24910?

File: templates/shared/codeowner_validation_messages.tmpl
Comment (JakobDev) [inline templates/shared/codeowner_validation_messages.tmpl:1]: You should just set `FileError`. This is already implemented in Gitea, so you don't need this file.

**Comment by smophat7** [comment]
> Duplicate of #24910?

Yeah. Looks like it follows the GitHub CODEOWNERS syntax rules though, and it requires write access to be an owner. Shield icon on the files is helpful.

**Comment by bcallaghan-et** [comment]
It's been a while since anyone on my team has looked at this. If any further changes are required to the CODEOWNERS feature (in either our implementation or the official implementation), it would be easier for all parties to create a new issue and work fresh from the latest version of Gitea.

**Pull request closed without merging**